### PR TITLE
MGMT-7114: support CPUArchitecture in Cluster

### PR DESCRIPTION
**Pull request by danielerez**
# Assisted Pull Request

## Description

As a preparation to support arm64 CPU architecture, introduced the following changes:
-  swagger: Added a 'cpu_architecture' property to the following objects:
    - cluster/cluster-create-params
    - infra-env/infra-env-create-params
-  Added cpuArchitecture to GetOpenshiftVersion func in versions:
   - Try to fetch a release image url and version from ReleaseImages list according to the requested CPU architecture.
   - An empty cpu_architecture implies the default x86_64.
- When registering a new cluster, set CPUArchitecture according to parms.
  - Validate by fetching available architectures (versions -> GetCPUArchitectures). 
- Notes:
  - Day2 flow supports only x86_64 for now.
  - Will be updated in following PRs: handle_ocp_versions.py/generate ISO flow/upload ISO flow.
  - This PR is based on changes in https://github.com/openshift/assisted-service/pull/2420

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

As this PR is adding an infrastructure for later arm64 support, no substantial impact in behaviour should occur. I.e. added tests should cover the changes.

## Assignees

/cc @YuviGold
/cc @filanov
/cc @osherdp
/cc @avishayt
/cc @yevgeny-shnaidman 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by filanov** [comment]
> Day2 flow supports only x86_64 for now.

Will this be handled in a different PR?

File: internal/common/test_configuration.go [hunk 1]
```diff
@@ -54,6 +55,7 @@ var (
 	ReleaseImage            = "quay.io/openshift-release-dev/ocp-release:4.6.16-x86_64"
 	RhcosImage              = "rhcos_4.6.0"
 	RhcosVersion            = "version-46.123-0"
+	CPUArchitecture         = "x86_64"
```
Comment (filanov) [inline internal/common/test_configuration.go:58]:
```suggestion
	CPUArchitecture         = DefaultCPUArchitecture
```

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -629,6 +636,31 @@ func verifyMinimalOpenShiftVersionForSingleNode(requestedOpenshiftVersion string
...
+	if err != nil {
+		return "", err
+	}
+	for _, cpuArchitecture := range cpuArchitectures {
+		if cpuArchitecture == newClusterParams.CPUArchitecture {
+			return cpuArchitecture, nil
+		}
+	}
+
+	if newClusterParams.CPUArchitecture != common.DefaultCPUArchitecture {
```
Comment (filanov) [inline internal/bminventory/inventory.go:656]: which case does this `if` covers? 
is there a case that `newClusterParams.CPUArchitecture` is not inside `cpuArchitectures` and we still don't return an error?

**Review by filanov — COMMENTED** [review]
Just making sure but is there a ticket for v2? 
cc: @nmagnezi it will probably affect your code changes as well.

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (danielerez) [reply internal/bminventory/inventory.go:656]: This validation is for a case in which a non-default architecture is requested, but wasn't found in available architectures list. So, it is not relevant for default architecture as x86_64 is available regardless (for backwards compatibility) - returned in line 661.

**Comment by danielerez** [comment]
> > Day2 flow supports only x86_64 for now.
> 
> Will this be handled in a different PR?

As day2 won't be handled as part of the arm64 effort, this PR is already handling RegisterAddHostsClusterInternal to use only the default architecture (x86_64).

File: internal/common/test_configuration.go (hunk 1 above)
Comment (danielerez) [reply internal/common/test_configuration.go:58]: right, done.

**Comment by filanov** [comment]
> > > Day2 flow supports only x86_64 for now.
> > 
> > 
> > Will this be handled in a different PR?
> 
> As day2 won't be handled as part of the arm64 effort, this PR is already handling RegisterAddHostsClusterInternal to use only the default architecture (x86_64).

OK but we still will need it eventually no? lest at least have a ticket for that.

**Comment by danielerez** [comment]
> > > > Day2 flow supports only x86_64 for now.
> > > 
> > > 
> > > Will this be handled in a different PR?
> > 
> > 
> > As day2 won't be handled as part of the arm64 effort, this PR is already handling RegisterAddHostsClusterInternal to use only the default architecture (x86_64).
> 
> OK but we still will need it eventually no? lest at least have a ticket for that.

Not as part of the current arm64 effort - we'll support SNO only (just none platform is available for now...).

**Comment by filanov** [comment]
> > > > > Day2 flow supports only x86_64 for now.
> > > > 
> > > > 
> > > > Will this be handled in a different PR?
> > > 
> > > 
> > > As day2 won't be handled as part of the arm64 effort, this PR is already handling RegisterAddHostsClusterInternal to use only the default architecture (x86_64).
> > 
> > 
> > OK but we still will need it eventually no? lest at least have a ticket for that.
> 
> Not as part of the current arm64 effort - we'll support SNO only (just none platform is available for now...).

i'm always forget that it's relevant mostly for SNO, thanks

**Comment by filanov** [comment]
LGTM from me, prefer to have another person to take a look.

**Comment by danielerez** [comment]
> Just making sure but is there a ticket for v2?
> cc: @nmagnezi it will probably affect your code changes as well.

Added the CPUArchitecture property also for infraenv in this PR. So v2 should be handled here (at least regarding infraenv).

**Comment by yevgeny-shnaidman** [comment]
General question: cpu architecture of infra env and cluster must be the same? Do we have validations for it? During host register, for example?

**Comment by danielerez** [comment]
> General question: cpu architecture of infra env and cluster must be the same? Do we have validations for it? During host register, for example?

Yeah the architecture must be identical for the cluster and infraenv. This PR is actually not handling late binding support, so no such validation yet. Does it make more sense to add that validation to BindHost rather than RegisterHost (as BindHost must have a cluster iiuc)?

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -629,6 +636,33 @@ func verifyMinimalOpenShiftVersionForSingleNode(requestedOpenshiftVersion string
 	}
 	return nil
 }
+
+func (b *bareMetalInventory) getNewClusterCPUArchitecture(newClusterParams *models.ClusterCreateParams) (string, error) {
+	if newClusterParams.CPUArchitecture == "" {
```
[the remark this answers is not in the document]
Comment (danielerez) [reply internal/bminventory/inventory.go:641]: Good idea, would be cleaner. Just need to modify the comments, as we probably want to remove these checks once using only release images list (which might not contain the default arch).

**Pull request merged**